### PR TITLE
feat(traces): add reranking span kind for document reranking in llama index

### LINKED ISSUE #1153: [trace][semantic] attributes for re-ranking
Use-cases:
Sometimes the re
Use cohere re-ranker
Use GPT based prompt re-ranker
General cross-encoder based re-ranker

- [ ] model or strategy used to re-rank
- [ ] final re-ordering of documents
- [ ] scores from the model
- [ ] re-ranked documents

This will allow us to compute NCDG of retrieval

### PR DESCRIPTION
**Pull request by RogerHYang**
resolves #1153 

https://github.com/Arize-ai/phoenix/assets/80478925/af999cdb-1de0-4ad3-8c16-b4606d800f7c

### CODE REVIEW

File: app/src/pages/trace/TracePage.tsx
Comment (axiomofjoy) [inline app/src/pages/trace/TracePage.tsx:595]: Strict equality? `===`

File: app/src/pages/trace/TracePage.tsx
Comment (axiomofjoy) [inline app/src/pages/trace/TracePage.tsx:609]: Same as above.

**Review by axiomofjoy — COMMENTED** [review]
Python code lgtm. Any reason to keep the test you added as an integration test, or can it be made into a unit test?

Let's get a review from @mikeldking for the frontend.

**Comment by axiomofjoy** [comment]
Looks awesome @RogerHYang !

File: app/src/pages/trace/TracePage.tsx
Comment (mikeldking) [reply app/src/pages/trace/TracePage.tsx:595]: this is okay because it encompasses null and undefined

File: app/schema.graphql
Comment (mikeldking) [inline app/schema.graphql:579]: minor nit that we already have `retriever` - it might make more sense for this to be `reranker` to keep in the same naming family
```
reranker = LLMRerank(choice_batch_size=5, top_n=reranker_top_n, service_context=service_context)
```

Comment (mikeldking) [reply app/schema.graphql:579]: https://python.langchain.com/docs/integrations/retrievers/cohere-reranker

**Comment by mikeldking** [comment]
we need a followup for langchain too

**Comment by mikeldking** [comment]
Screenshot 2023-10-10 at 5 29 56 PM
The LLM reranker of Llama_index doesn't get the same attributes as the CohereReranker - I think if we are going to calculate things like NCDG we will have to at least capture the attributes from both. Do you know what the root cause of his is?

File: app/src/pages/trace/TracePage.tsx
Comment (mikeldking) [inline app/src/pages/trace/TracePage.tsx:624]: there is a `titleExtra` prop on card where you can place a `Counter` component. https://5f9739a76e154c00220dd4b9-zeknbennzf.chromatic.com/?path=/story/counter--gallery

File: app/src/pages/trace/TracePage.tsx
Comment (mikeldking) [inline app/src/pages/trace/TracePage.tsx:649]: Same as above - rely on titleExtra and counter.

File: app/src/pages/trace/TracePage.tsx
Comment (mikeldking) [inline app/src/pages/trace/TracePage.tsx:643]: curious if this is largely not useful if as you say as the retrieval ordering is not useful - or is the idea to show the `score` of the original retrieval alongside the final rank-ordered one? I think if it's just duplication of data it might make more sense to just nix this.

File: app/src/pages/trace/TracePage.tsx
Comment (mikeldking) [inline app/src/pages/trace/TracePage.tsx:640]: I might re-color these just so that there's a visual hierarchy of color (e.g. that re-ranked documents take on a different tint) - this way as you are clicking around you can clearly see the difference.

Comment (mikeldking) [reply app/src/pages/trace/TracePage.tsx:640]: Re-using the `DocumentItem` component is good but I think just showing the new `score` label might be a tad confusing? Or just using `score` as an abstract is intended here.

File: integration-tests/trace/llama_index/test_callback.py
Comment (mikeldking) [inline integration-tests/trace/llama_index/test_callback.py:44]: quick question on this parameter: https://docs.cohere.com/docs/reranking

I'm guessing this is the same as TOP_N? If you feed say 5 documents but pass top_k of 3, does it only rank 3? Just trying to understand why this is a parameter to the rerank model.

**Review by mikeldking — COMMENTED** [review]
LGTM! I filed some follow-up tickets. Might do a quick audit on the UI post any changes but looks good!

File: app/src/pages/trace/TracePage.tsx
Comment (mikeldking) [reply app/src/pages/trace/TracePage.tsx:640]: In one case it's often spacial distance where as when running through a reranker it is a relevance rank. Just thinking that from a user's perspective displaying `score: XX` alongside both we lose a bit of an opportunity to explain the `score` in this context a bit better - `score` being pretty generic.

File: integration-tests/trace/llama_index/test_callback.py
Comment (RogerHYang) [reply integration-tests/trace/llama_index/test_callback.py:44]: Yes, this is the same as TOP_N. (The letter is `K` in [literature](https://nlp.stanford.edu/IR-book/html/htmledition/inexact-top-k-document-retrieval-1.html) because `N` is usually the total number of docs.) The caller of the reranker usually just wants a relatively small number of docs out of a initial set of tens or hundreds. It's certainly optional because it can just rank each document, but in general, a reduction in number is expected for each stage of the retrieval process.

File: app/src/pages/trace/TracePage.tsx
Comment (RogerHYang) [reply app/src/pages/trace/TracePage.tsx:640]: I think even though `score` is generic, it is still accurate. On the input side of the reranker, `score` may or may not exist, and even if it does exist, it's not considered by the reranker. But if the "input" `score` does exist it was generated by a preprocessor for a separate purpose. The general mental picture here is that there could be millions of documents in a corpus, and only a relatively small set are chosen to be reranked, and that selection process can have a `score` of its own based on the query in question. Even though that `score` is not meaningful to the reranker, it is still an informative attribute of the input document, because it relays the reason for how the document became a candidate in the first place (especially when the preprocessor is missing in the trace). On the other hand, we can't really get more specific that the `score` verbiage because we don't have more information. On balance, although it may seem confusing at first, a user should have enough context to reason their way through it.

File: app/src/pages/trace/TracePage.tsx
Comment (RogerHYang) [reply app/src/pages/trace/TracePage.tsx]: I added it and it looks like this.

Screenshot 2023-10-12 at 8 59 14 AM

File: app/schema.graphql
Comment (RogerHYang) [reply app/schema.graphql]: it's been renamed

File: app/src/pages/trace/TracePage.tsx
Comment (RogerHYang) [reply app/src/pages/trace/TracePage.tsx:644]: Are you considering not including this in the semantic convention at all, or just not displaying it in the UI? If it's already included in the span then displaying it seems a natural thing to do, even though the ordering is not meaningful.

File: integration-tests/trace/llama_index/test_callback.py
Comment (mikeldking) [reply integration-tests/trace/llama_index/test_callback.py]: Still confused though - if I pass say 5 documents with `top_k` of 3 - does it rank 5 and trim the last two?

File: tutorials/internal/llama_index_tracing_example.ipynb
Comment (mikeldking) [inline tutorials/internal/llama_index_tracing_example.ipynb]: this is the internal one - it seemed nice to have a notebook that executed a reranker?

File: app/src/pages/trace/TracePage.tsx
Comment (mikeldking) [reply app/src/pages/trace/TracePage.tsx:644]: I think it might largely be duplicate to the retriever but yeah we can leave it for now for transparency reasons. Bummer on the ordering. I think there will be asks about ordering but we can wait for feedback.

File: app/src/pages/trace/TracePage.tsx
Comment (mikeldking) [reply app/src/pages/trace/TracePage.tsx:641]: I think I wasn't disputing the way we capture the score - was just thinking of ways to avoid the mental "eason their way through it." a bit. But I don't have an immediate good prefix for the reranker score so let's keep it for now.

File: integration-tests/trace/llama_index/test_callback.py
Comment (RogerHYang) [reply integration-tests/trace/llama_index/test_callback.py]: Yes, it's retains up to `K` in the output, so in this case 2 (of the lowest scores) has to be dropped. `Top K` is applied after ranking all 5 docs.

File: tutorials/internal/llama_index_tracing_example.ipynb
Comment (RogerHYang) [reply tutorials/internal/llama_index_tracing_example.ipynb]: oh, my bad. I had mistaken this one with the public one `llama_index_tracing_tutorial.ipynb` which has a very similar name.

Comment (RogerHYang) [reply tutorials/internal/llama_index_tracing_example.ipynb]: reverted

File: app/src/pages/trace/TracePage.tsx
Comment (RogerHYang) [reply app/src/pages/trace/TracePage.tsx:644]: I think it's just that ordering is not guaranteed in the general case, but should exist in most cases. On the other hand, if the `score` is missing for any reason (since it's not essential to the reranker), the ordering would be lost regardless, so it's also not reliable.

**Review by mikeldking — APPROVED** [review]

**Pull request merged**
